app/src/main/java/de/westnordost/streetcomplete/data/osmnotes/edits/NoteEditsUploader.kt

```diff
@@ -55,17 +60,33 @@ class NoteEditsUploader @Inject constructor(
     }
 
     private fun uploadEdit(edit: NoteEdit) {
-        val text = edit.text.orEmpty() + uploadAndGetAttachedPhotosText(edit.imagePaths)
 
+        // try to upload the image
+        var imageText = edit.uploadedDataMap.getOrDefault("images", "")
+        if(imageText.isEmpty()) {
+            imageText = uploadAndGetAttachedPhotosText(edit.imagePaths)
+            noteEditsController.updateData(edit, "images", imageText)
+        }
+
+        // try to upload the tracks
+        var tracksText = edit.uploadedDataMap.getOrDefault("tracks", "")
+        if(tracksText.isEmpty()) {
+            tracksText = uploadAndGetAttachedTracksText(edit.tracks, edit.text)
+            noteEditsController.updateData(edit, "tracks", tracksText)
+        }
```

Comment:
Okay, I understand now what this "uploadedDataMap" is for. It enables that if one step of the possibly two steps prior to actually creating the note (due to a network problem) fails, the same images / tracks are not uploaded again on next try

That's fine, though not really important or actually a little problematic because
1. uploaded photos not referenced by any notes will be cleaned up quite quickly
2. a few duplicately uploaded traces in case of rare upload connectivity problems don't really make a difference. GPX tracks are tiny in size
3. current implementation is not perfect anyway, because there may be several photos for upload and the upload may fail anytime while uploading the multiple photos
4. Because of point 1, upload of photos and creation of note must happen promptly after one another. Otherwise, the final note will just contain dead links (or worse, image activation fails, not sure what happens then)
So question is if it is really worth the added complexity.

However, if you opt to keep this logic:
1. I'd very much prefer if this is not a "nameless" JSON map but just two columns in the table: `uploadedImagesText` and `uploadedTracksText`
2. What if there are no images to attach / no tracks to attach? Then, the upload functions and the updateData function will be called regardless

Reply:
After pondering over the weekend, I have to agree with you that this isn't that important / necessary.  If there isn't any issue with duplicate images then my original concern is invalid. I think just removing the logic and making things simpler is the way to go.